test-data/unit/check-functions.test

```diff
@@ -733,6 +733,94 @@ a = None # type: A
 a.f()
 a.f(None) # E: Too many arguments for "f" of "A"
 
+[case testUntypedDecoratorWithDeclaredType]
+
+from typing import Callable
+from mypy_extensions import decorated_type
+
+def dec(f): pass
+
+@decorated_type(Callable[[int], str])
+@dec
+def f(): pass
+
+f("a")  # E: Argument 1 to "f" has incompatible type "str"; expected "int"
+x: str = f(1)
+y: int = f(1)  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+
+reveal_type(f)  # E: Revealed type is 'def (builtins.int) -> builtins.str'
+
+[builtins fixtures/dict.pyi]
+
+[case testDecoratorWithDeclaredTypeBare]
+
+from typing import Callable
+from mypy_extensions import decorated_type
+
+@decorated_type(Callable[[int], str])
+def f(x): pass
+
+f("a")  # E: Argument 1 to "f" has incompatible type "str"; expected "int"
+x: str = f(1)
+y: int = f(1)  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+reveal_type(f)  # E: Revealed type is 'def (builtins.int) -> builtins.str'
+
+[builtins fixtures/dict.pyi]
+
+[case testDecoratorWithDeclaredTypeIncompatible]
+
+from typing import Callable, TypeVar
+from mypy_extensions import decorated_type
+
+T = TypeVar('T')
+
+def dec(f: T) -> T: pass
+
+@decorated_type(Callable[[int], str])  # E: Incompatible types (inferred decorated type Callable[[str], str], declared decorated type Callable[[int], str])
```

Comment:
Add test where the declared type is a proper supertype of the inferred type. It should be okay -- i.e. test that the subtype checks goes the right way.